src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -208,7 +234,38 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
                 }
                 if (customBuilder)
                 {
-                    taskProperty = GetCustomTaskProperty(F, builderType);
+                    builderType = builderArgument as NamedTypeSymbol;
+
+                    // builderType must be an open generic of arity 1, and its containing types must be nongeneric
+                    bool isBuilderGenericityOk = ((object)builderType != null &&
+                                                  !builderType.IsErrorType() &&
+                                                  builderType.Arity == 1 && 
+                                                  builderType.TypeArguments[0] is UnboundArgumentErrorTypeSymbol);
+
+                    for (var containingType = builderType.ContainingType;
+                        containingType != null && isBuilderGenericityOk;
+                        containingType = containingType.ContainingType)
+                    {
+                        isBuilderGenericityOk &= (containingType.Arity == 0);
+                    }
+
+                    if (builderType?.IsErrorType() == true)
+                    {
+                    }
+                    else if ((object)builderType == null ||
+                        builderType.SpecialType == SpecialType.System_Void ||
+                        builderType.DeclaredAccessibility != returnType.DeclaredAccessibility ||
+                        !isBuilderGenericityOk)
+                    {
```

Comment:
Fixed in https://github.com/dotnet/roslyn/pull/13405/commits/eef0c70b2966698aa71be1063a920290e5778f51#diff-995d1bf33206eef8c980dbd08b8601b0R268